File: TranslationFiestaGo/internal/utils/logger.go
package utils

import (
	"fmt"
	"log"
	"os"
	"path/filepath"
	"time"
)

// LogLevel represents the severity level of a log message
type LogLevel int

const (
	DEBUG LogLevel = iota
	INFO
	WARN
	ERROR
)

// Logger provides structured logging functionality
type Logger struct {
	level  LogLevel
	logger *log.Logger
	file   *os.File
}

// Global logger instance
var globalLogger *Logger

// NewLogger creates a new logger instance
func NewLogger(level LogLevel, logFile string) (*Logger, error) {
	var file *os.File
	var err error

	if logFile != "" {
		// Create log directory if it doesn't exist
		logDir := filepath.Dir(logFile)
		if err := os.MkdirAll(logDir, 0755); err != nil {
			return nil, fmt.Errorf("failed to create log directory: %w", err)
		}

		file, err = os.OpenFile(logFile, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
		if err != nil {
			return nil, fmt.Errorf("failed to open log file: %w", err)
		}
	}

	logger := &Logger{
		level:  level,
		logger: log.New(file, "", log.LstdFlags),
		file:   file,
	}

	return logger, nil
}

// GetLogger returns the global logger instance
func GetLogger() *Logger {
	return globalLogger
}

// InitLogger initializes the global logger
func InitLogger(level LogLevel, logFile string) error {
	logger, err := NewLogger(level, logFile)
	if err != nil {
		return err
	}

	globalLogger = logger
	return nil
}

// Close closes the logger and its file handle
func (l *Logger) Close() error {
	if l.file != nil {
		return l.file.Close()
	}
	return nil
}

// log writes a message to the log with the specified level
func (l *Logger) log(level LogLevel, format string, args ...interface{}) {
	if level < l.level {
		return
	}

	levelStr := l.getLevelString(level)
	timestamp := time.Now().Format("2006-01-02 15:04:05")

	message := fmt.Sprintf(format, args...)
	logEntry := fmt.Sprintf("[%s] %s: %s", timestamp, levelStr, message)

	if l.logger != nil {
		l.logger.Println(logEntry)
	} else {
		fmt.Println(logEntry)
	}
}

// Debug logs a debug message
func (l *Logger) Debug(format string, args ...interface{}) {
	l.log(DEBUG, format, args...)
}

// Info logs an info message
func (l *Logger) Info(format string, args ...interface{}) {
	l.log(INFO, format, args...)
}

// Warn logs a warning message
func (l *Logger) Warn(format string, args ...interface{}) {
	l.log(WARN, format, args...)
}

// Error logs an error message
func (l *Logger) Error(format string, args ...interface{}) {
	l.log(ERROR, format, args...)
}

// getLevelString returns the string representation of a log level
func (l *Logger) getLevelString(level LogLevel) string {
	switch level {
	case DEBUG:
		return "DEBUG"
	case INFO:
		return "INFO"
	case WARN:
		return "WARN"
	case ERROR:
		return "ERROR"
	default:
		return "UNKNOWN"
	}
}

// Convenience functions for global logger
func Debug(format string, args ...interface{}) {
	if globalLogger != nil {
		globalLogger.Debug(format, args...)
	}
}

func Info(format string, args ...interface{}) {
	if globalLogger != nil {
		globalLogger.Info(format, args...)
	}
}

func Warn(format string, args ...interface{}) {
	if globalLogger != nil {
		globalLogger.Warn(format, args...)
	}
}

func Error(format string, args ...interface{}) {
	if globalLogger != nil {
		globalLogger.Error(format, args...)
	}
}

File: flutter_translate/lib/core/utils/logger.dart
/// Clean Code logging system with thread-safe operations
/// Following Single Responsibility principle and meaningful naming
library;

import 'dart:io';
import 'dart:async';
import 'package:path_provider/path_provider.dart';
import 'package:intl/intl.dart';

/// Log levels following Clean Code naming conventions
enum LogLevel {
  debug('DEBUG'),
  info('INFO'),
  warning('WARN'),
  error('ERROR');

  const LogLevel(this.value);
  final String value;
}

/// Thread-safe logger implementation with file output
/// Single Responsibility: Handle all logging operations
class Logger {
  static Logger? _instance;
  static const String _logFileName = 'flutter_translate.log';
  static final DateFormat _dateFormat = DateFormat('yyyy-MM-dd HH:mm:ss.SSS');

  late File _logFile;
  final StreamController<String> _logController =
      StreamController<String>.broadcast();
  bool _isInitialized = false;

  /// Private constructor for singleton pattern
  Logger._();

  /// Get the singleton instance
  static Logger get instance {
    _instance ??= Logger._();
    return _instance!;
  }

  /// Initialize the logger with file output
  Future<void> initialize() async {
    if (_isInitialized) return;

    try {
      final directory = await getApplicationDocumentsDirectory();
      final logPath = '${directory.path}/$_logFileName';
      _logFile = File(logPath);

      // Ensure log file exists
      if (!await _logFile.exists()) {
        await _logFile.create(recursive: true);
      }

      _isInitialized = true;
      info('Logger initialized successfully');
    } catch (e) {
      // Fallback to stdout if file logging fails
      stderr.writeln('Failed to initialize logger: $e');
      _isInitialized = false;
    }
  }

  /// Log a debug message
  void debug(String message) => _log(LogLevel.debug, message);

  /// Log an info message
  void info(String message) => _log(LogLevel.info, message);

  /// Log a warning message
  void warning(String message) => _log(LogLevel.warning, message);

  /// Log an error message
  void error(String message) => _log(LogLevel.error, message);

  /// Internal logging method with thread-safe file operations
  void _log(LogLevel level, String message) {
    final timestamp = _dateFormat.format(DateTime.now().toUtc());
    final logEntry = '[$timestamp] ${level.value}: $message\n';

    // Add to stream for real-time updates if needed
    _logController.add(logEntry);

    if (!_isInitialized) {
      // Fallback to console if not initialized
      print(logEntry.trim());
      return;
    }

    // Write to file asynchronously and safely
    _writeToFile(logEntry);
  }

  /// Thread-safe file writing operation
  void _writeToFile(String logEntry) async {
    try {
      await _logFile.writeAsString(
        logEntry,
        mode: FileMode.append,
        flush: true,
      );
    } catch (e) {
      // Best effort logging - don't crash the app
      stderr.writeln('Failed to write to log file: $e');
    }
  }

  /// Get the current log file path
  String? get logFilePath => _isInitialized ? _logFile.path : null;

  /// Stream of log entries for real-time monitoring
  Stream<String> get logStream => _logController.stream;

  /// Clean up resources
  void dispose() {
    _logController.close();
  }

  /// Get recent log entries (useful for debugging)
  Future<List<String>> getRecentLogs({int maxLines = 100}) async {
    if (!_isInitialized) return [];

    try {
      final content = await _logFile.readAsString();
      final lines =
          content.split('\n').where((line) => line.trim().isNotEmpty).toList();

      return lines.length > maxLines
          ? lines.sublist(lines.length - maxLines)
          : lines;
    } catch (e) {
      error('Failed to read recent logs: $e');
      return [];
    }
  }

  /// Clear the log file (useful for testing or user request)
  Future<void> clearLogs() async {
    if (!_isInitialized) return;

    try {
      await _logFile.writeAsString('', mode: FileMode.write);
      info('Log file cleared');
    } catch (e) {
      error('Failed to clear log file: $e');
    }
  }
}

/// Convenience functions for easy logging
void logDebug(String message) => Logger.instance.debug(message);
void logInfo(String message) => Logger.instance.info(message);
void logWarning(String message) => Logger.instance.warning(message);
void logError(String message) => Logger.instance.error(message);
